include/unifex/heap_allocate.hpp

```diff
@@ -0,0 +1,113 @@
+
+#include <memory>
+#include <type_traits>
+
+namespace unifex {
+
+template <typename Operation, typename Allocator>
+class heap_allocated_operation {
+  using allocator_t = typename std::allocator_traits<Allocator>::template rebind_alloc<Operation>;
+
+public:
+  template<typename Sender, typename Receiver>
+  explicit heap_allocated_operation(Sender&& s, Receiver&& r)
+      : allocator_(get_allocator(r)) {
+    using allocator_traits = std::allocator_traits<allocator_t>;
+    Operation* op = allocator_traits::allocate(allocator_, 1);
+    bool constructorSucceeded = false;
+    scope_guard freeOnError = [&]() noexcept {
+      if (!constructorSucceeded) {
+        allocator_traits::deallocate(allocator_, op, 1);
+      }
+    };
+    op_ = ::new (static_cast<void*>(op)) Operation(
+      cpo::connect((Sender&&)s, (Receiver&&)r));
+    constructorSucceeded = true;
+  }
+
+  ~heap_allocated_operation() {
+    op_->~Operation();
+    std::allocator_traits<allocator_t>::deallocate(allocator_, op_, 1);
+  }
+
+  void start() noexcept {
+      cpo::start(*op_);
+  }
+
+private:
+  Operation *op_;
+  UNIFEX_NO_UNIQUE_ADDRESS allocator_t allocator_;
+};
+
+template <typename Sender>
+class heap_allocate_sender {
+public:
+  template<
+    template<typename...> class Variant,
+    template<typename...> class Tuple>
+  using value_types = typename Sender::template value_types<Variant, Tuple>;
+
+  template<template<typename...> class Variant>
+  using error_types = typename Sender::template error_types<Variant>;
+
+  template <typename Receiver>
+  friend auto tag_invoke(
+      tag_t<cpo::connect>, heap_allocate_sender&& s, Receiver&& r)
+      -> heap_allocated_operation<operation_t<Sender, Receiver>,
+                                  std::remove_cvref_t<get_allocator_t<Receiver>>> {
+    return heap_allocated_operation<operation_t<Sender, Receiver>,
+                                  std::remove_cvref_t<get_allocator_t<Receiver>>>{
+        (Sender&&)s.sender_, (Receiver&&) r};
+  }
+
+  template <typename Receiver>
+  friend auto tag_invoke(
+      tag_t<cpo::connect>, heap_allocate_sender& s, Receiver&& r)
+      -> heap_allocated_operation<operation_t<Sender&, Receiver>,
+                                  std::remove_cvref_t<get_allocator_t<Receiver>>> {
+    return heap_allocated_operation<operation_t<Sender&, Receiver>,
+                                  std::remove_cvref_t<get_allocator_t<Receiver>>>{
+        s.sender_, (Receiver&&) r};
+  }
+
+  template <typename Receiver>
+  friend auto tag_invoke(
+      tag_t<cpo::connect>, const heap_allocate_sender& s, Receiver&& r)
+      -> heap_allocated_operation<operation_t<const Sender&, Receiver>,
+                                  std::remove_cvref_t<get_allocator_t<Receiver>>> {
+    return heap_allocated_operation<operation_t<const Sender&, Receiver>,
+                                  std::remove_cvref_t<get_allocator_t<Receiver>>>{
+        std::as_const(s.sender_), (Receiver&&) r};
```

Comment:
`as_const` is not needed here, is it?

Reply:
Yes, it's not strictly needed as long as you assume that the `Sender` template is not a non-const lvalue reference.